### ISSUE TITLE
Issues when rewriting conflicts opam2

### ISSUE THREAD

**Issue by samoht** [ocaml/opam#2958]
```
$ opam install git.1.11.0 irmin.0.10.1
Your request can't be satisfied:
  - git.1.11.0 is in conflict with irmin.0.10.1

$ opam2 install git.1.11.0 irmin.0.10.1
The following actions will be performed:
  ∗  install uchar                   0.0.1    [required by fmt]
  ∗  install dolog                   3.0      [required by irmin]
  ∗  install astring                 0.8.3    [required by git]
  ∗  install ocamlgraph              1.8.7    [required by git, irmin]
  ∗  install base-bytes              base     [required by uri, decompress, mirage-tc]
  ∗  install jbuilder                1.0+beta9  [required by git]
  ∗  install ppx_tools               5.0      [required by lwt]
  ∗  install base-num                base     [required by num]
  ∗  install crunch                  2.0.0    [required by irmin]
  ∗  install uutf                    1.0.1    [required by jsonm]
  ∗  install fmt                     0.8.3    [required by git]
  ∗  install stringext               1.4.3    [required by irmin]
  ∗  install re                      1.7.1    [required by irmin]
  ∗  install decompress              0.6      [required by git]
  ∗  install cppo                    1.5.0    [required by ocplib-endian]
  ∗  install ppx_traverse_builtins   v0.9.0   [required by ppx_core, ppx_metaquot]
  ∗  install ocaml-migrate-parsetree 1.0.1    [required by ppx_sexp_conv, bin_prot]
  ∗  install ocaml-compiler-libs     v0.9.0   [required by ppx_core]
  ∗  install lwt                     2.6.0    [required by git, irmin]
  ∗  install num                     0        [required by sexplib]
  ∗  install jsonm                   1.0.1    [required by ezjsonm]
  ∗  install ocplib-endian           1.0      [required by cstruct]
  ∗  install ppx_ast                 v0.9.1   [required by ppx_core]
  ∗  install logs                    0.6.2    [required by git]
  ∗  install sexplib                 v0.9.1   [required by uri, ezjsonm]
  ∗  install cstruct                 3.0.0    [required by irmin]
  ∗  install base                    v0.9.2   [required by bin_prot]
  ∗  install mstruct                 1.3.3    [required by git, irmin]
  ∗  install hex                     1.1.1    [required by git, irmin]
  ∗  install stdio                   v0.9.0   [required by ppx_core]
  ∗  install ezjsonm                 0.4.3    [required by irmin]
  ∗  install ppx_core                v0.9.0   [required by ppx_sexp_conv]
  ∗  install ppx_optcomp             v0.9.0   [required by ppx_driver]
  ∗  install ppx_driver              v0.9.1   [required by ppx_sexp_conv, bin_prot]
  ∗  install variantslib             v0.9.0   [required by ppx_variants_conv]
  ∗  install ppx_metaquot            v0.9.0   [required by ppx_sexp_conv]
  ∗  install fieldslib               v0.9.0   [required by ppx_fields_conv]
  ∗  install ppx_type_conv           v0.9.0   [required by ppx_sexp_conv]
  ∗  install ppx_variants_conv       v0.9.0   [required by bin_prot]
  ∗  install ppx_traverse            v0.9.0   [required by ppx_custom_printf]
  ∗  install ppx_sexp_conv           v0.9.0   [required by uri]
  ∗  install ppx_fields_conv         v0.9.0   [required by bin_prot]
  ∗  install ppx_compare             v0.9.0   [required by bin_prot]
  ∗  install uri                     1.9.4    [required by git, irmin]
  ∗  install ppx_custom_printf       v0.9.0   [required by bin_prot]
  ∗  install git                     1.11.0
  ∗  install bin_prot                v0.9.0   [required by mirage-tc]
  ∗  install mirage-tc               0.3.0    [required by irmin]
  ∗  install irmin                   0.10.1
===== ∗  49 =====
Do you want to continue ? [Y/n]
```

Could it be related to the `&` in the conflict clause of `irmin.0.10.1`?:
```
conflicts: [
  "git" {< "1.7.1" & > "1.7.2"}
  ...
```

**Comment by samoht** [comment]
I've pushed https://github.com/ocaml/opam-repository/pull/9425/commits/59b815940896975797cc21c997047194721aa649 to try to fix the conflict rewriting.

**Comment by AltGr** [comment]
See #2267 and #2246; there is indeed a difference in semantics in 2.0, because that of 1.2 was unsound.

Remember that `&` means intersection, so the only meaningful interpretation of `< "1.7.1" & > "1.7.2"` is `<empty>`. opam 1.2 forces you to write `&` connectors, but then treats the results as a disjunction. I think I checked that no existing packages from opam-repository used that when I made the change (and didn't expect the next release to take >2 years at the time).

Note that writing `"git" {< "1.7.1" > "1.7.2"}` will be interpreted as you expect in both cases. I'm not sure wether rewriting `&` to `|` would be a good move: we keep feature-compatibility, but by accepting someting broken on 1.2, which will lead users to errors after the migration.

Of course, `lint` could check for empty ranges in package formulas...
